airflow/executors/celery_executor.py

```diff
@@ -274,6 +330,61 @@ def execute_async(self,
     def terminate(self):
         pass
 
+    def try_adopt_task_instances(self, tis: List[TaskInstance]) -> List[TaskInstance]:
+        # See which if the tis are still alive (or have finished even!)
+        #
+        # Since Celery doesn't store "SENT" state for queued commands (if we create an AsyncResult with a made
+        # up id it just returns PENDING state for it), we have to store Celery's task_id against the TI row to
+        # look at in future.
+        #
+        # This process is not perfect -- we could have sent the task to celery, and crashed before we were
+        # able to record the AsyncResult.task_id in the TaskInstance table, in which case we won't adopt the
+        # task (it'll either run and update the TI state, or the scheduler will clear and re-queue it. Either
+        # way it won't get executed more than once)
+        #
+        # (If we swapped it around, and generated a task_id for Celery, stored that in TI and enqueued that
+        # there is also still a race condition where we could generate and store the task_id, but die before
+        # we managed to enqueue the command. Since neither way is perfect we always have to deal with this
+        # process not being perfect.)
+
+        celery_tasks = {
+            ti.external_executor_id: AsyncResult(ti.external_executor_id)
+            for ti in tis
+            if ti.external_executor_id is not None
+        }
+
+        if not celery_tasks:
+            # Nothing to adopt
+            return tis
+
+        tis_by_celery_task_id = {ti.external_executor_id: ti for ti in tis}
+
+        states_by_celery_task_id = self.bulk_state_fetcher.get_many(celery_tasks.values())
+
+        not_adopted = [ti for ti in tis if ti.external_executor_id is None]
+        adopted = []
+        cached_celery_backend = next(iter(celery_tasks.values())).backend
+
+        for celery_task_id, (state, info) in states_by_celery_task_id.items():
+            ti = tis_by_celery_task_id[celery_task_id]
+            result = celery_tasks[celery_task_id]
+            result.backend = cached_celery_backend
```

Comment:
Maybe we don't need to have `tis_by_celery_task_id = {ti.external_executor_id: ti for ti in tis}` (line 360) separately, instead, we can have 
```python
celery_tasks = {
    ti.external_executor_id: (AsyncResult(ti.external_executor_id), ti)
    for ti in tis
    if ti.external_executor_id is not None
}
```

Then lines 369-370 can also be merged into `result, ti = celery_tasks[celery_task_id]`

By doing this, we can avoid traversing `tis` two times, also save one query on the dict.

Reply:
In addition, `tis_by_celery_task_id` covers all `tis`, which is not necessary, because it will only be used to query `tis` which present in `celery_tasks` (essentially means `tis` whose `external_executor_id` is not `None`).

So my suggested revision above may save a little memory as well ;-)

Reply:
Nice!

Reply:
Done.